### ISSUE TITLE
Calendar System

### ISSUE THREAD

**Issue by elthran** [elthran/BookingSystem#9]
I want to fully outline the Calendar here. Then each part of it we might even break down into further issues. My current calendar on our website looks like this (purely for testing purposes, there is no CSS or JavaScript yet): 
![calendar](https://user-images.githubusercontent.com/11773722/41097579-0da5f4a0-6a8b-11e8-8777-3beddd2d834b.jpg)

I am just setting it up so you can click on any given day of the month and then on the right side of the page will show available booking times. This is not at all how it will look when we are done, but I needed something to test my database on. Let me know exactly what you want it to do (but necessarily how it will look). Like for example the JaneApp would be:

- You Choose a Treatment first and the Calendar loads to the very first day where you can book that treatment
- You can click a button at the top to bring up a calendar menu and choose any day of the year
- There is another button at the top which lets you move ahead/behind one day at a time
- On the page itself, it lists each employee separately and for each employee it lists each available appointment time as a button (in 1 hour increments which further splits into 30 minute increments if selected, and im not sure if I like that aspect of it)
- Clicking on a time secures that time for 10 minutes and offers you a link to fill in you data and permanently book it
- On the side of the page there are descriptions of other treatments and clicking on one changes that current treatment as well as jumps you to the next available day for that treatment
- There is a link to return to the home page of the business booking but not many other links

Those seem to be the functions of their calendar page. The layout makes no difference, but I need to know how similar you want our page so I can do a mock-up. This will likely take a week or two as there are many functions to add and some are complicated.

**Comment by melissachang9** [comment]
Choosing a practitioner and then listing their services or vice versa is fine.  

I like having the time slots as buttons on the calendar however I agree that I don't want it to then show a pop up screen where time is show in half hour increments.  This wouldn't be applicable to my business.  Just show available time slots on the calendar as buttons for the treatment and practitioner selected.

Everything else you've mentioned is good.  Keep in mind, we will need to have calendars attached to practitioners at different locations as well.

**Comment by elthran** [comment]
We will need to discuss "practitioners at different locations". I will need to fully understand what we are aiming for before I can start coding it. I assume that means a single business might have multiple "calendar objects" and each employee/client is linked to only one calendar. So it would go
- Business -> 
- Calendar (each calendar is linked to one business but businesses have multiple calendars) ->
- Practitioners (each practitioner is linked to one business and one calendar, but businesses and calendars have multiple practitioners each) ->
- Clients (each business/calendar/practitioner has multiple clients. But client linking back is more complex. In my code each client can only be linked to one business (ie. if that same client books at a new business, a new client profile is created for that business). But I don't think you want the same for calendars/practitioners. So I will need to rethink how this relationship works and maybe rework my database a bit.

Does that sound right?

If that's all right it shouldn't be too hard, I will just keep clients linked to one business and probably just have each "appointment" linked to a certain practitioner/calendar. So a client has no relationship to calendars/practitioners, just a business, and the appointments are what are linked. Then everything is one-to-many relationship and very easy to query.

**Comment by melissachang9** [comment]
Yes I think that sounds correct. To give you a scenario, let’s say my business I have a location in Vancouver but also one in Nanaimo. Certain employees work at certain locations or both and can be booked accordingly. So if I’m working in Nanaimo on Thursday, then a client is able to book me there and then book me for Friday in Vancouver.  So practitioners shouldn’t be tied to one location. 

Another side note, also important to ensure that different services are available on different days. So although I may offer massage on Mondays, it’s not available on Tuesdays even if I’m available for different services on Tuesday. 

Melissa Chang

**Comment by elthran** [comment]
> Another side note, also important to ensure that different services are available on different days. So although I may offer massage on Mondays, it’s not available on Tuesdays even if I’m available for different services on Tuesday.

This part is easy. Each service will have its own calendar and they will overlay onto the final calendar.

For clients, are they separate for each business location at all? Or will the business just have one list of clients that is shared? Let me know how you want that, as I'm still trying to decide how to code in Locations. Right now the site assumes each business has one location, butI will add those in this week.

**Comment by melissachang9** [comment]
Ok so the user will be able to choose which services to provide on certain days and locations.

One list of clients for each business.


Melissa Chang

**Comment by elthran** [comment]
Ok that's simple. Would there be any sort of location at all tied to the client? Like so you can only list clients that went to this location? Or just 100% the business has a list of clients and no mention of location (much simpler and how it currently is).

**Comment by melissachang9** [comment]
No location tied to clients

Melissa Chang
